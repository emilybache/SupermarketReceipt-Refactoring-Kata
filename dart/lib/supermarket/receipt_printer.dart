import 'package:supermarket_receipt/supermarket/model/product_unit.dart';
import 'package:supermarket_receipt/supermarket/model/receipt.dart';
import 'package:supermarket_receipt/supermarket/model/receipt_item.dart';
import 'package:supermarket_receipt/supermarket/model/discount.dart';

class ReceiptPrinter {
  
  final int _columns;

  ReceiptPrinter() : this._columns = 40;

  ReceiptPrinter.withColumns(this._columns);

  String printReceipt(Receipt receipt) {
    StringBuffer result = StringBuffer();
    for (ReceiptItem item in receipt.getItems()) {
      String receiptItem = presentReceiptItem(item);
      result.write(receiptItem);
    }
    for (Discount discount in receipt.getDiscounts()) {
      String discountPresentation = presentDiscount(discount);
      result.write(discountPresentation);
    }

    result.write('\n');
    result.write(presentTotal(receipt));
    return result.toString();
  }

  String presentReceiptItem(ReceiptItem item) {
    String totalPricePresentation = presentPrice(item.getTotalPrice());
    String name = item.getProduct().getName();

    String line = formatLineWithWhitespace(name, totalPricePresentation);

    if (item.getQuantity() != 1) {
      line += '  ' + presentPrice(item.getPrice()) + ' * ' + presentQuantity(item) + '\n';
    }
    return line;
  }

  String presentDiscount(Discount discount) {
    String name = discount.getDescription() + ' (' + discount.getProduct().getName() + ')';
    String value = presentPrice(discount.getDiscountAmount());

    return formatLineWithWhitespace(name, value);
  }

  String presentTotal(Receipt receipt) {
    String name = 'Total: ';
    String value = presentPrice(receipt.getTotalPrice());
    return formatLineWithWhitespace(name, value);
  }

  String formatLineWithWhitespace(String name, String value) {
    StringBuffer line = StringBuffer();
    line.write(name);
    int whitespaceSize = _columns - name.length - value.length;
    for (int i = 0; i < whitespaceSize; i++) {
      line.write(' ');
    }
    line.write(value);
    line.write('\n');
    return line.toString();
  }

  String presentPrice(double price) {
    return price.toStringAsFixed(2);
  }

  String presentQuantity(ReceiptItem item) {
    if (item.getProduct().getUnit() == ProductUnit.EACH) {
      return item.getQuantity().toInt().toString();
    } else {
      return item.getQuantity().toStringAsFixed(3);
    }
  }
}