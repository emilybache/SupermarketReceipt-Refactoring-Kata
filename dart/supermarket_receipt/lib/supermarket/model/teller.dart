import 'package:supermarket_receipt/supermarket/model/offer.dart';
import 'package:supermarket_receipt/supermarket/model/product.dart';
import 'package:supermarket_receipt/supermarket/model/product_quantity.dart';
import 'package:supermarket_receipt/supermarket/model/receipt.dart';
import 'package:supermarket_receipt/supermarket/model/shopping_cart.dart';
import 'package:supermarket_receipt/supermarket/model/special_offer_type.dart';
import 'package:supermarket_receipt/supermarket/model/supermarket_catalog.dart';

class Teller {

  final SupermarketCatalog _catalog;
  final Map<Product, Offer> _offers = {};

  Teller(this._catalog);

  void addSpecialOffer(SpecialOfferType offerType, Product product, double argument) {
    _offers[product] = Offer(offerType, product, argument);
  }

  Receipt checksOutArticlesFrom(ShoppingCart theCart) {
    Receipt receipt = Receipt();
    List<ProductQuantity> productQuantities = theCart.getItems();
    for (var pq in productQuantities) {
      Product p = pq.getProduct();
      double quantity = pq.getQuantity();
      double unitPrice = _catalog.getUnitPrice(p);
      double price = quantity * unitPrice;
      receipt.addProduct(p, quantity, unitPrice, price);
    }
    theCart.handleOffers(receipt, _offers, _catalog);

    return receipt;
  }
}