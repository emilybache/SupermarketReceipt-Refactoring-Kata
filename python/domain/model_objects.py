from enum import Enum


class Product:
    def __init__(self, name, unit):
        self.name = name
        self.unit = unit


class ProductQuantity:
    def __init__(self, product, quantity):
        self.product = product
        self.quantity = quantity


class ProductUnit(Enum):
    EACH = 1
    KILO = 2


class SpecialOfferType(Enum):
    THREE_FOR_TWO = 1
    TEN_PERCENT_DISCOUNT = 2
    TWO_FOR_AMOUNT = 3
    FIVE_FOR_AMOUNT = 4


class Offer:
    def __init__(self, offer_type, product, argument):
        self.offer_type = offer_type
        self.product = product
        self.argument = argument


class Discount:
    def __init__(self, product, description, discount_amount):
        self.product = product
        self.description = description
        self.discount_amount = discount_amount
