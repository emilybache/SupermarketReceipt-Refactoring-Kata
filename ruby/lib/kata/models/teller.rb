class Kata::Teller

  def initialize(catalog)
    @catalog = catalog
    @offers = {}
  end

  def add_special_offer(offer_type, product, argument)
    @offers[product] = Kata::Offer.new(offer_type, product, argument)
  end

  def checks_out_articles_from(the_cart)
    receipt = Kata::Receipt.new
    product_quantities = the_cart.items
    for pq in product_quantities do
      p = pq.product
      quantity = pq.quantity
      unit_price = @catalog.unit_price(p)
      price = quantity * unit_price
      receipt.add_product(p, quantity, unit_price, price)
    end
    the_cart.handle_offers(receipt, @offers, @catalog)

    receipt
  end

end
