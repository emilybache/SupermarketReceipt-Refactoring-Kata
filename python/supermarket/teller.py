from supermarket.catalog import SupermarketCatalog
from typing import Any, Optional

from supermarket.shopping_cart import ShoppingCart
from supermarket.product import ProductName, Product, ProductQuantity
from supermarket.offer import create_offer, Offer, SpecialOfferType
from supermarket.receipt import Receipt


class Teller:
    def __init__(self, catalog: SupermarketCatalog) -> None:
        self.catalog = catalog
        self.offers: dict[ProductName, Offer] = {}

    def add_special_offer(
        self,
        offer_type: SpecialOfferType,
        product: Product,
        argument: Optional[Any] = None,
    ) -> None:
        print(f"creating offer {offer_type}")
        self.offers[product.name] = create_offer(offer_type, product, argument)

    def checks_out_articles_from(self, cart: ShoppingCart) -> Receipt:
        receipt = Receipt()
        pq: ProductQuantity
        for pq in cart.items:
            product = pq.product
            quantity = pq.quantity
            unit_price: float = self.catalog.unit_price(product.name)
            price = quantity * unit_price
            receipt.add_product(product, quantity, unit_price, price)

        self.handle_offers(receipt, cart)

        return receipt

    def handle_offers(self, receipt: Receipt, cart: ShoppingCart):
        for product_name, quantity in cart.product_quantities.items():
            if product_name in self.offers.keys():
                offer = self.offers[product_name]
                unit_price = self.catalog.unit_price(product_name)
                discount = offer.calculate_discount(unit_price, quantity)
                if discount:
                    receipt.add_discount(discount)
