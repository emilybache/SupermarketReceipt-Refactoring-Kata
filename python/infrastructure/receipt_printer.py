from python.domain.model_objects import ProductUnit


class ReceiptPrinter:
    """
    Service responsible for formatting and printing receipts.

    This printer handles the text-based receipt format.
    Future implementations could include HTML, PDF, or other formats.
    """

    def __init__(self, columns=40):
        self.columns = columns

    def print_receipt(self, receipt):
        """
        Generate a text representation of the receipt.

        Args:
            receipt: Receipt object containing items and discounts

        Returns:
            str: Formatted receipt text
        """
        result = ""

        # Add all items
        for item in receipt.items:
            receipt_item = self._format_receipt_item(item)
            result += receipt_item

        # Add all discounts
        for discount in receipt.discounts:
            discount_line = self._format_discount(discount)
            result += discount_line

        # Add total
        result += "\n"
        result += self._format_total(receipt)

        return str(result)

    def _format_receipt_item(self, item):
        """Format a single receipt item"""
        total_price_printed = self._format_price(item.total_price)
        name = item.product.name
        line = self._format_line_with_whitespace(name, total_price_printed)

        if item.quantity != 1:
            quantity_info = f"  {self._format_price(item.price)} * {self._format_quantity(item)}\n"
            line += quantity_info

        return line

    def _format_line_with_whitespace(self, name, value):
        """Format a line with proper spacing between name and value"""
        line = name
        whitespace_size = self.columns - len(name) - len(value)

        # Ensure minimum spacing
        if whitespace_size < 1:
            whitespace_size = 1

        line += " " * whitespace_size
        line += value
        line += "\n"
        return line

    def _format_price(self, price):
        """Format a price value"""
        return "%.2f" % price

    def _format_quantity(self, item):
        """Format quantity based on product unit type"""
        if ProductUnit.EACH == item.product.unit:
            return str(int(item.quantity)) if item.quantity == int(item.quantity) else str(item.quantity)
        else:
            return '%.3f' % item.quantity

    def _format_discount(self, discount):
        """Format a discount line"""
        name = f"{discount.description} ({discount.product.name})"
        value = self._format_price(discount.discount_amount)
        return self._format_line_with_whitespace(name, value)

    def _format_total(self, receipt):
        """Format the total line"""
        name = "Total: "
        value = self._format_price(receipt.total_price())
        return self._format_line_with_whitespace(name, value)
