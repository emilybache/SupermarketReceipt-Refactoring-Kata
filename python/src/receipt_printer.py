from model_objects import ProductUnit

class ReceiptPrinter:

    def __init__(self, columns=40):
        self.columns = columns
  
    def print_receipt(self, receipt):
        result = ""
        for item in receipt.items:
            receipt_item = self.print_receipt_item(item)
            result += receipt_item

        for discount in receipt.discounts:
            discount_presentation = self.print_discount(discount)
            result += discount_presentation

        result += "\n"
        result += self.present_total(receipt)
        return str(result)

    def print_receipt_item(self, item):
        total_price_printed = self.print_price(item.total_price)
        name = item.product.name
        line = self.format_line_with_whitespace(name, total_price_printed)
        if item.quantity != 1:
            line += f"  {self.print_price(item.price)} * {self.print_quantity(item)}\n"
        return line

    def format_line_with_whitespace(self, name, value):
        line = name
        whitespace_size = self.columns - len(name) - len(value)
        for i in range(whitespace_size):
            line += " "
        line += value
        line += "\n"
        return line

    def print_price(self, price):
        return "%.2f" % price

    def print_quantity(self, item):
        if ProductUnit.EACH == item.product.unit:
            return str(item.quantity)
        else:
            return '%.3f' % item.quantity

    def print_discount(self, discount):
        name = f"{discount.description} ({discount.product.name})"
        value = self.print_price(discount.discount_amount)
        return self.format_line_with_whitespace(name, value)

    def present_total(self, receipt):
        name = "Total: "
        value = self.print_price(receipt.total_price())
        return self.format_line_with_whitespace(name, value)
