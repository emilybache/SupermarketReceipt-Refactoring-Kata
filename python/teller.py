from model_objects import Offer
from receipt import Receipt


class Teller:

    def __init__(self, catalog):
        self.catalog = catalog
        self.offers = {}

    def add_special_offer(self, offer_type, product, argument):
        self.offers[product] = Offer(offer_type, product, argument)

    def checks_out_articles_from(self, the_cart):
        receipt = Receipt()
        product_quantities = the_cart.items
        for pq in product_quantities:
            p = pq.product
            quantity = pq.quantity
            unit_price = self.catalog.unit_price(p)
            price = quantity * unit_price
            receipt.add_product(p, quantity, unit_price, price)

        the_cart.handle_offers(receipt, self.offers, self.catalog)

        return receipt
