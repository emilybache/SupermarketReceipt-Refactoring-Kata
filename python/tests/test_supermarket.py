import approvaltests

from model_objects import SpecialOfferType
from receipt_printer import ReceiptPrinter
from teller import Teller


def test_basic_functionality(catalog_with_toothbrush_and_apple, apple_product, shopping_cart):
    teller = Teller(catalog_with_toothbrush_and_apple)
    shopping_cart.add_item_quantity(apple_product, 2.5)

    receipt = teller.checks_out_articles_from(shopping_cart)

    string_receipt = ReceiptPrinter().print_receipt(receipt)

    approvaltests.verify(string_receipt)


def test_ten_percent_discount(catalog_with_toothbrush, toothbrush_product, shopping_cart):
    teller = Teller(catalog_with_toothbrush)
    teller.add_special_offer(SpecialOfferType.TEN_PERCENT_DISCOUNT, toothbrush_product, 10.0)

    shopping_cart.add_item_quantity(toothbrush_product, 2.5)
    receipt = teller.checks_out_articles_from(shopping_cart)
    string_receipt = ReceiptPrinter().print_receipt(receipt)

    approvaltests.verify(string_receipt)


def test_three_for_two_discount(catalog_with_toothbrush, toothbrush_product, shopping_cart):
    teller = Teller(catalog_with_toothbrush)
    teller.add_special_offer(SpecialOfferType.THREE_FOR_TWO, toothbrush_product, 2)

    shopping_cart.add_item_quantity(toothbrush_product, 3)
    receipt = teller.checks_out_articles_from(shopping_cart)
    string_receipt = ReceiptPrinter().print_receipt(receipt)

    approvaltests.verify(string_receipt)


def test_two_items_for_reduced_price_discount(catalog_with_toothbrush, toothbrush_product, shopping_cart):
    teller = Teller(catalog_with_toothbrush)
    teller.add_special_offer(SpecialOfferType.TWO_FOR_AMOUNT, toothbrush_product, 0.99)

    shopping_cart.add_item_quantity(toothbrush_product, 2)
    receipt = teller.checks_out_articles_from(shopping_cart)
    string_receipt = ReceiptPrinter().print_receipt(receipt)

    approvaltests.verify(string_receipt)


def test_five_items_for_reduced_price_discount(catalog_with_toothbrush, toothbrush_product, shopping_cart):
    teller = Teller(catalog_with_toothbrush)
    teller.add_special_offer(SpecialOfferType.FIVE_FOR_AMOUNT, toothbrush_product, 4)

    shopping_cart.add_item_quantity(toothbrush_product, 9)
    receipt = teller.checks_out_articles_from(shopping_cart)
    string_receipt = ReceiptPrinter().print_receipt(receipt)

    approvaltests.verify(string_receipt)

