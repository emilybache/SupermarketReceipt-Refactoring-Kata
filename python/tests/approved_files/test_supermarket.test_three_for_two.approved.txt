toothbrush                          2.97
  0.99 * 3
3 for 2 (toothbrush)               -0.99

Total:                              1.98
